server/src/main/java/com/linecorp/centraldogma/server/command/StandaloneCommandExecutor.java

```diff
@@ -195,10 +241,48 @@ protected void doStop(@Nullable Runnable onReleaseLeadership) {
     }
 
     private CompletableFuture<Revision> push(PushCommand c) {
+        if (c.projectName().equals(INTERNAL_PROJ) || c.repositoryName().equals(Project.REPO_DOGMA) ||
+            Double.compare(permitsPerSecond, -1) == 0) {
+            return push0(c);
+        }
+
+        return getRateLimiter(c.projectName(), c.repositoryName()).thenCompose(rateLimiter -> {
+            if (rateLimiter == null || rateLimiter.tryAcquire()) {
+                return push0(c);
+            } else {
+                return CompletableFutures.exceptionallyCompletedFuture(
+                        new TooManyRequestsException("commits", c.executionPath(), rateLimiter.getRate()));
+            }
+        });
+    }
+
+    private CompletableFuture<Revision> push0(PushCommand c) {
         return repo(c).commit(c.baseRevision(), c.timestamp(),
                               c.author(), c.summary(), c.detail(), c.markup(), c.changes());
     }
 
+    private CompletableFuture<RateLimiter> getRateLimiter(String projectName, String repoName) {
+        return metadataService.getRepo(projectName, repoName).thenApply(meta -> {
+            final QuotaConfig writeQuota = meta.writeQuota();
+            final double permitsForRepo = writeQuota != null ? writeQuota.permitsPerSecond() : 0;
+            final double permitsPerSecond =
+                    permitsForRepo != 0 ? permitsForRepo : this.permitsPerSecond;
+            if (permitsPerSecond > 0) {
+                return writeRateLimiters.compute(projectName + '/' + repoName, (key, rateLimiter) -> {
```

Comment:
How about adding todo for removing the rate limiter when the repository is removed?

Reply:
I can address it in this PR. :-)